pkg/render/policyrecommendation.go

```diff
@@ -0,0 +1,413 @@
+}
+
+func (pr *policyRecommendationComponent) deploymentPodTemplate() *corev1.PodTemplateSpec {
+	var ps []corev1.LocalObjectReference
+	for _, x := range pr.cfg.PullSecrets {
+		ps = append(ps, corev1.LocalObjectReference{Name: x.Name})
+	}
+
+	volumes := []corev1.Volume{
+		pr.cfg.TrustedBundle.Volume(),
+	}
+
+	container := relasticsearch.ContainerDecorateIndexCreator(
+		relasticsearch.ContainerDecorate(pr.controllerContainer(), pr.cfg.ESClusterConfig.ClusterName(),
+			ElasticsearchPolicyRecommendationUserSecret, pr.cfg.ClusterDomain, rmeta.OSTypeLinux),
+		pr.cfg.ESClusterConfig.Replicas(), pr.cfg.ESClusterConfig.Shards())
+
+	return relasticsearch.DecorateAnnotations(&corev1.PodTemplateSpec{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        PolicyRecommendationName,
+			Namespace:   PolicyRecommendationNamespace,
+			Annotations: pr.intrusionDetectionAnnotations(),
+		},
+		Spec: corev1.PodSpec{
+			Tolerations:        pr.cfg.Installation.ControlPlaneTolerations,
+			NodeSelector:       pr.cfg.Installation.ControlPlaneNodeSelector,
+			ServiceAccountName: PolicyRecommendationName,
+			ImagePullSecrets:   ps,
+			Containers: []corev1.Container{
+				container,
+			},
+			Volumes: volumes,
+		},
+	}, pr.cfg.ESClusterConfig, pr.cfg.ESSecrets).(*corev1.PodTemplateSpec)
+}
+
+func (pr *policyRecommendationComponent) intrusionDetectionAnnotations() map[string]string {
+	return pr.cfg.TrustedBundle.HashAnnotations()
+}
+
+func (pr *policyRecommendationComponent) serviceAccount() client.Object {
+	return &corev1.ServiceAccount{
+		TypeMeta:   metav1.TypeMeta{Kind: "ServiceAccount", APIVersion: "v1"},
+		ObjectMeta: metav1.ObjectMeta{Name: PolicyRecommendationServiceAccountName, Namespace: PolicyRecommendationNamespace},
+	}
+}
+
+// allowTigeraPolicyForPolicyRecommendation defines an allow-tigera policy for policy recommendation.
+func allowTigeraPolicyForPolicyRecommendation(cfg *PolicyRecommendationConfiguration) *v3.NetworkPolicy {
+	kubeSystemEntityRule := networkpolicy.CreateEntityRule("kube-system", "kube-dns", 53)
+	apiserverEntityRule := v3.EntityRule{
+		NamespaceSelector: "projectcalico.org/name == 'default'",
+		Selector: "provider == 'kubernetes' && component == 'apiserver' && " +
+			"endpoints.projectcalico.org/serviceName == 'kubernetes'",
+		Ports: networkpolicy.Ports(443, 6443, 12388),
+	}
+
+	egressRules := []v3.Rule{
+		{
+			Action:      v3.Allow,
+			Protocol:    &networkpolicy.TCPProtocol,
+			Destination: apiserverEntityRule,
+		},
+		{
+			Action:      v3.Allow,
+			Protocol:    &networkpolicy.TCPProtocol,
+			Destination: networkpolicy.ESGatewayEntityRule,
+		},
+		{
+			Action:      v3.Allow,
+			Protocol:    &networkpolicy.TCPProtocol,
+			Destination: DexEntityRule,
+		},
+	}
+
+	if !cfg.Openshift {
+		egressRules = append(egressRules, v3.Rule{
+			Action:      v3.Allow,
+			Protocol:    &networkpolicy.UDPProtocol,
+			Destination: kubeSystemEntityRule,
+		})
+	} else {
+		egressRules = append(egressRules, v3.Rule{
```

Comment:
there is a util function for the DNS rules in `networkpolicy.AppendDNSEgressRules`